File: packages/agentenkit/src/adapters/docker.ts
import { spawn } from 'node:child_process';
import {
  DEFAULT_COMMAND_TIMEOUT_MS,
  SandboxFileNotFoundError,
  SandboxGoneError,
  SandboxUnsupportedError,
  TIMED_OUT_EXIT_CODE,
  type CommandResult,
  type CreateSandboxOptions,
  type FileEntry,
  type RunCommandOptions,
  type Sandbox,
  type SandboxFileSystem,
  type SandboxInfo,
  type SandboxNetwork,
  type SandboxProvider,
} from '../ports/sandbox.js';
import {
  FILE_SCRIPTS,
  NOT_FOUND_EXIT,
  OutputCollector,
  backgroundCommand,
  looksTimedOut,
  parseListing,
  randomId,
  resolveIn,
  runArgv,
  timeoutSeconds,
  toBytes,
} from './sandbox-shell.js';

export interface DockerSandboxOptions {
  /** The image each sandbox runs. It needs `sh`; bash, Python and
   *  `timeout` are used when there. Default `python:3.12-slim`. */
  image?: string;
  /** `none` (the default) or `all`. Docker has no allow list by domain. */
  network?: 'none' | 'all';
  /** The folder commands start in. Default `/workspace`. */
  workdir?: string;
  cpu?: number;
  memoryMiB?: number;
  /** Ports to publish on 127.0.0.1, so `getUrl` can give them. Needs
   *  `network: 'all'`. */
  ports?: number[];
  /** Environment variables every command sees. */
  env?: Record<string, string>;
  /** More `docker run` arguments, placed before the image. */
  runArgs?: string[];
  /** The docker CLI. Default `docker` on the PATH. */
  docker?: string;
}

const DEADLINE_FILE = '/tmp/.agentenkit-deadline';

/** The container's main process: it waits until the deadline in
 *  DEADLINE_FILE has passed and then exits, and `--rm` removes the
 *  container. So a sandbox nobody uses ends by itself, even when the app
 *  that made it is gone. `setTimeout` moves the deadline. */
const WATCHDOG =
  `echo "$1" > ${DEADLINE_FILE}; ` +
  `while [ "$(date +%s)" -lt "$(cat ${DEADLINE_FILE} 2>/dev/null || echo 0)" ]; do sleep 2; done`;

/** What the docker CLI says when the container is not there any more. */
const GONE = /Error response from daemon: (No such container|container \S+ is not running)|No such object/;

/** Sandboxes as Docker containers on the machine the docker CLI talks to:
 *  for self-hosting. One container per thread, removed when it has been idle
 *  past its time. Uses the docker CLI, so it needs no SDK and follows your
 *  DOCKER_HOST and contexts. The Go runtime has the same adapter
 *  (adapters/docker). */
export class DockerSandbox implements SandboxProvider {
  readonly name = 'docker';
  private readonly docker: string;

  constructor(private readonly options: DockerSandboxOptions = {}) {
    this.docker = options.docker ?? 'docker';
  }

  async create(options: CreateSandboxOptions): Promise<Sandbox> {
    const network: SandboxNetwork = options.network ?? this.options.network ?? 'none';
    if (typeof network === 'object') throw new SandboxUnsupportedError('docker', 'a network allow list');
    const name = `agentenkit-${randomId()}`;
    const workdir = this.options.workdir ?? '/workspace';
    const deadline = Math.floor((Date.now() + options.timeoutMs) / 1000);
    const args = [
      'run', '-d', '--rm', '--init', '--name', name,
      '--label', 'agentenkit=sandbox',
      '--label', `agentenkit.threadId=${options.metadata.threadId}`,
      ...(options.metadata.runId ? ['--label', `agentenkit.runId=${options.metadata.runId}`] : []),
      '-w', workdir,
      '--network', network === 'all' ? 'bridge' : 'none',
      '--security-opt', 'no-new-privileges',
    ];
    const cpu = options.resources?.cpu ?? this.options.cpu;
    const memory = options.resources?.memoryMiB ?? this.options.memoryMiB;
    if (cpu) args.push('--cpus', String(cpu));
    if (memory) args.push('--memory', `${memory}m`);
    for (const port of this.options.ports ?? []) args.push('-p', `127.0.0.1::${port}`);
    for (const [k, v] of Object.entries({ ...this.options.env, ...options.envs })) args.push('-e', `${k}=${v}`);
    args.push(...(this.options.runArgs ?? []));
    args.push(options.image ?? this.options.image ?? 'python:3.12-slim', 'sh', '-c', WATCHDOG, 'sandbox', String(deadline));
    const r = await cli(this.docker, args);
    if (r.code !== 0) throw new Error(`docker: could not start a sandbox: ${text(r.stderr).trim()}`);
    return new DockerBox(this.docker, name, workdir);
  }

  async connect(sandboxId: string): Promise<Sandbox> {
    const r = await cli(this.docker, ['inspect', '-f', '{{.State.Running}}\t{{.Config.WorkingDir}}', sandboxId]);
    const [running, workdir] = text(r.stdout).trim().split('\t');
    if (r.code !== 0 || running !== 'true') throw new SandboxGoneError(sandboxId);
    return new DockerBox(this.docker, sandboxId, workdir || '/');
  }
}

class DockerBox implements Sandbox {
  readonly provider = 'docker';
  readonly filesystem: SandboxFileSystem;

  constructor(
    private readonly docker: string,
    readonly sandboxId: string,
    private readonly workdir: string,
  ) {
    this.filesystem = new DockerFiles(this);
  }

  /** `docker exec` a script with `sh -c`; a container that is gone throws. */
  async script(script: string, args: string[], stdin?: Uint8Array): Promise<Exec> {
    const r = await cli(this.docker, ['exec', ...(stdin ? ['-i'] : []), this.sandboxId, 'sh', '-c', script, 'sandbox', ...args], { stdin });
    this.checkGone(r);
    return r;
  }

  checkGone(r: Exec) {
    if (r.code !== 0 && GONE.test(text(r.stderr))) throw new SandboxGoneError(this.sandboxId);
  }

  path(p: string) {
    return resolveIn(this.workdir, p);
  }

  async runCommand(command: string, options: RunCommandOptions = {}): Promise<CommandResult> {
    const limit = options.timeoutMs ?? DEFAULT_COMMAND_TIMEOUT_MS;
    const args = ['exec', '-w', this.path(options.cwd ?? '.')];
    for (const [k, v] of Object.entries(options.env ?? {})) args.push('-e', `${k}=${v}`);
    const cmd = options.background ? backgroundCommand(command) : command;
    args.push(this.sandboxId, ...runArgv(cmd, options.background ? undefined : timeoutSeconds(limit)));
    const started = Date.now();
    const out = new OutputCollector(options.background ? undefined : options.onStdout);
    const err = new OutputCollector(options.background ? undefined : options.onStderr);
    // `timeout` inside the container stops the command; this is the backstop
    // for an image without it.
    const r = await cli(this.docker, args, { out, err, limitMs: limit + 5_000, signal: options.signal });
    const stdout = out.end();
    const stderr = err.end();
    if (r.code !== 0 && GONE.test(stderr)) throw new SandboxGoneError(this.sandboxId);
    const durationMs = Date.now() - started;
    if (options.background) return { stdout: '', stderr: '', exitCode: r.code, durationMs };
    const timedOut = r.killed || looksTimedOut(r.code, durationMs, limit);
    return {
      stdout,
      stderr,
      exitCode: timedOut ? TIMED_OUT_EXIT_CODE : r.code,
      durationMs,
      ...(timedOut ? { timedOut: true } : {}),
      ...(out.truncated || err.truncated ? { truncated: true } : {}),
    };
  }

  async getUrl(options: { port: number; protocol?: 'http' | 'https' }): Promise<string> {
    const r = await cli(this.docker, ['port', this.sandboxId, `${options.port}/tcp`]);
    this.checkGone(r);
    const hostPort = text(r.stdout).trim().split('\n')[0];
    if (r.code !== 0 || !hostPort) {
      throw new SandboxUnsupportedError('docker', `a URL for port ${options.port} (publish it with the ports option)`);
    }
    return `${options.protocol ?? 'http'}://${hostPort.replace('0.0.0.0', '127.0.0.1')}`;
  }

  async getInfo(): Promise<SandboxInfo> {
    const r = await cli(this.docker, [
      'inspect', '-f', '{{.Created}}\t{{.State.Status}}\t{{json .Config.Labels}}', this.sandboxId,
    ]);
    if (r.code !== 0) throw new SandboxGoneError(this.sandboxId);
    const [created, status, labels] = text(r.stdout).trim().split('\t');
    const metadata: Record<string, string> = {};
    for (const [k, v] of Object.entries((JSON.parse(labels || '{}') ?? {}) as Record<string, string>)) {
      if (k.startsWith('agentenkit.')) metadata[k.slice('agentenkit.'.length)] = v;
    }
    const deadline = Number(text((await this.script(`cat ${DEADLINE_FILE}`, [])).stdout).trim());
    return {
      id: this.sandboxId,
      provider: 'docker',
      status: status === 'running' ? 'running' : status === 'exited' ? 'stopped' : 'error',
      createdAt: new Date(created!),
      ...(deadline ? { expiresAt: new Date(deadline * 1000) } : {}),
      workdir: this.workdir,
      metadata,
    };
  }

  async setTimeout(timeoutMs: number): Promise<void> {
    const deadline = Math.floor((Date.now() + timeoutMs) / 1000);
    const r = await this.script(`echo "$1" > ${DEADLINE_FILE}`, [String(deadline)]);
    if (r.code !== 0) throw new Error(`docker: could not set the sandbox's time: ${text(r.stderr).trim()}`);
  }

  async destroy(): Promise<void> {
    const r = await cli(this.docker, ['rm', '-f', this.sandboxId]);
    if (r.code !== 0 && !GONE.test(text(r.stderr))) {
      throw new Error(`docker: could not remove ${this.sandboxId}: ${text(r.stderr).trim()}`);
    }
  }
}

class DockerFiles implements SandboxFileSystem {
  constructor(private readonly box: DockerBox) {}

  private async run(script: string, path: string, stdin?: Uint8Array): Promise<Exec> {
    const r = await this.box.script(script, [this.box.path(path)], stdin);
    if (r.code === NOT_FOUND_EXIT) throw new SandboxFileNotFoundError(path);
    if (r.code !== 0) throw new Error(`docker: ${text(r.stderr).trim() || `exit ${r.code}`}`);
    return r;
  }

  async readFile(path: string): Promise<string> {
    return text((await this.run(FILE_SCRIPTS.read, path)).stdout);
  }

  async readFileBytes(path: string): Promise<Uint8Array> {
    return (await this.run(FILE_SCRIPTS.read, path)).stdout;
  }

  async writeFile(path: string, content: string | Uint8Array): Promise<void> {
    await this.run(FILE_SCRIPTS.write, path, toBytes(content));
  }

  async readdir(path: string): Promise<FileEntry[]> {
    return parseListing(text((await this.run(FILE_SCRIPTS.list, path)).stdout));
  }

  async mkdir(path: string): Promise<void> {
    await this.run(FILE_SCRIPTS.mkdir, path);
  }

  async exists(path: string): Promise<boolean> {
    return (await this.box.script(FILE_SCRIPTS.exists, [this.box.path(path)])).code === 0;
  }

  async remove(path: string): Promise<void> {
    await this.run(FILE_SCRIPTS.remove, path);
  }
}

interface Exec {
  code: number;
  stdout: Uint8Array;
  stderr: Uint8Array;
  /** Stopped by `limitMs`. */
  killed?: boolean;
}

const text = (b: Uint8Array) => new TextDecoder().decode(b);

/** Runs the docker CLI. With `out` / `err` the output goes there as it
 *  comes; otherwise it is collected whole. */
function cli(
  docker: string,
  args: string[],
  opts: { stdin?: Uint8Array; out?: OutputCollector; err?: OutputCollector; limitMs?: number; signal?: AbortSignal } = {},
): Promise<Exec> {
  return new Promise((resolve, reject) => {
    const child = spawn(docker, args, { stdio: [opts.stdin ? 'pipe' : 'ignore', 'pipe', 'pipe'] });
    const stdout: Buffer[] = [];
    const stderr: Buffer[] = [];
    child.stdout!.on('data', (b: Buffer) => (opts.out ? opts.out.push(b) : stdout.push(b)));
    child.stderr!.on('data', (b: Buffer) => (opts.err ? opts.err.push(b) : stderr.push(b)));
    let killed = false;
    let aborted = false;
    const timer = opts.limitMs
      ? setTimeout(() => {
          killed = true;
          child.kill('SIGKILL');
        }, opts.limitMs)
      : undefined;
    const onAbort = () => {
      aborted = true;
      child.kill('SIGKILL');
    };
    opts.signal?.addEventListener('abort', onAbort, { once: true });
    if (opts.signal?.aborted) onAbort();
    const done = () => {
      if (timer) clearTimeout(timer);
      opts.signal?.removeEventListener('abort', onAbort);
    };
    child.on('error', (e) => {
      done();
      reject(new Error(`docker: could not run ${docker}: ${e.message}`));
    });
    child.on('close', (code) => {
      done();
      if (aborted) return reject(opts.signal!.reason ?? new Error('aborted'));
      resolve({ code: code ?? -1, stdout: new Uint8Array(Buffer.concat(stdout)), stderr: new Uint8Array(Buffer.concat(stderr)), killed });
    });
    if (opts.stdin) child.stdin!.end(Buffer.from(opts.stdin));
  });
}

File: packages/go-agentenkit/adapters/docker/docker.go
// Package docker runs sandboxes as Docker containers on the machine the
// docker CLI talks to: for self-hosting. One container per thread, removed
// when it has been idle past its time. It uses the docker CLI, so it needs
// no SDK and follows your DOCKER_HOST and contexts. The TS runtime has the
// same adapter (DockerSandbox).
package docker

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/internal/sandboxsh"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// Options shapes the provider.
type Options struct {
	// Image is what each sandbox runs. It needs sh; bash, Python and
	// timeout are used when there. Default python:3.12-slim.
	Image string
	// Network is "none" (the default) or "all". Docker has no allow list by
	// domain.
	Network string
	// Workdir is the folder commands start in. Default /workspace.
	Workdir   string
	CPU       float64
	MemoryMiB int
	// Ports to publish on 127.0.0.1, so GetURL can give them. Needs
	// Network "all".
	Ports []int
	// Env holds environment variables every command sees.
	Env map[string]string
	// RunArgs are more docker run arguments, placed before the image.
	RunArgs []string
	// Docker is the docker CLI. Default docker on the PATH.
	Docker string
}

// Provider makes Docker sandboxes.
type Provider struct{ opts Options }

// New returns the provider.
func New(opts Options) *Provider {
	if opts.Docker == "" {
		opts.Docker = "docker"
	}
	return &Provider{opts: opts}
}

// Name implements ports.SandboxProvider.
func (p *Provider) Name() string { return "docker" }

const deadlineFile = "/tmp/.agentenkit-deadline"

// watchdog is the container's main process: it waits until the deadline in
// deadlineFile has passed and then exits, and --rm removes the container.
// So a sandbox nobody uses ends by itself, even when the app that made it
// is gone. SetTimeout moves the deadline.
const watchdog = `echo "$1" > ` + deadlineFile + `; ` +
	`while [ "$(date +%s)" -lt "$(cat ` + deadlineFile + ` 2>/dev/null || echo 0)" ]; do sleep 2; done`

// gone is what the docker CLI says when the container is not there any
// more.
var gone = regexp.MustCompile(`Error response from daemon: (No such container|container \S+ is not running)|No such object`)

// Create implements ports.SandboxProvider.
func (p *Provider) Create(ctx context.Context, opts ports.CreateSandboxOptions) (ports.Sandbox, error) {
	network := p.opts.Network
	if opts.Network != nil {
		network = opts.Network.Mode
	}
	if network == "allow" {
		return nil, ports.SandboxUnsupported("docker", "a network allow list")
	}
	name := "agentenkit-" + sandboxsh.RandomID(12)
	workdir := p.opts.Workdir
	if workdir == "" {
		workdir = "/workspace"
	}
	deadline := time.Now().Add(opts.Timeout).Unix()
	args := []string{
		"run", "-d", "--rm", "--init", "--name", name,
		"--label", "agentenkit=sandbox",
		"--label", "agentenkit.threadId=" + opts.Metadata.ThreadID,
	}
	if opts.Metadata.RunID != "" {
		args = append(args, "--label", "agentenkit.runId="+opts.Metadata.RunID)
	}
	netName := "none"
	if network == "all" {
		netName = "bridge"
	}
	args = append(args, "-w", workdir, "--network", netName, "--security-opt", "no-new-privileges")
	cpu, memory := p.opts.CPU, p.opts.MemoryMiB
	if opts.Resources.CPU > 0 {
		cpu = opts.Resources.CPU
	}
	if opts.Resources.MemoryMiB > 0 {
		memory = opts.Resources.MemoryMiB
	}
	if cpu > 0 {
		args = append(args, "--cpus", strconv.FormatFloat(cpu, 'f', -1, 64))
	}
	if memory > 0 {
		args = append(args, "--memory", fmt.Sprintf("%dm", memory))
	}
	for _, port := range p.opts.Ports {
		args = append(args, "-p", fmt.Sprintf("127.0.0.1::%d", port))
	}
	for _, env := range []map[string]string{p.opts.Env, opts.Envs} {
		for k, v := range env {
			args = append(args, "-e", k+"="+v)
		}
	}
	args = append(args, p.opts.RunArgs...)
	image := opts.Image
	if image == "" {
		image = p.opts.Image
	}
	if image == "" {
		image = "python:3.12-slim"
	}
	args = append(args, image, "sh", "-c", watchdog, "sandbox", strconv.FormatInt(deadline, 10))
	r, err := cli(ctx, p.opts.Docker, args, nil, nil, nil)
	if err != nil {
		return nil, err
	}
	if r.code != 0 {
		return nil, fmt.Errorf("docker: could not start a sandbox: %s", strings.TrimSpace(r.stderr.String()))
	}
	return p.box(name, workdir), nil
}

// Connect implements ports.SandboxProvider.
func (p *Provider) Connect(ctx context.Context, sandboxID string) (ports.Sandbox, error) {
	r, err := cli(ctx, p.opts.Docker, []string{"inspect", "-f", "{{.State.Running}}\t{{.Config.WorkingDir}}", sandboxID}, nil, nil, nil)
	if err != nil {
		return nil, err
	}
	fields := strings.SplitN(strings.TrimSpace(r.stdout.String()), "\t", 2)
	if r.code != 0 || fields[0] != "true" {
		return nil, ports.SandboxGone(sandboxID, "")
	}
	workdir := "/"
	if len(fields) == 2 && fields[1] != "" {
		workdir = fields[1]
	}
	return p.box(sandboxID, workdir), nil
}

func (p *Provider) box(id, workdir string) *box {
	b := &box{docker: p.opts.Docker, id: id, workdir: workdir}
	b.files = &files{b: b}
	return b
}

type box struct {
	docker, id, workdir string
	files               *files
}

func (b *box) ID() string                          { return b.id }
func (b *box) Provider() string                    { return "docker" }
func (b *box) Filesystem() ports.SandboxFileSystem { return b.files }
func (b *box) path(p string) string                { return sandboxsh.ResolveIn(b.workdir, p) }

// script runs a script with sh -c through docker exec; a container that is
// gone is an error.
func (b *box) script(ctx context.Context, script string, args []string, stdin []byte) (result, error) {
	argv := []string{"exec"}
	if stdin != nil {
		argv = append(argv, "-i")
	}
	argv = append(argv, b.id, "sh", "-c", script, "sandbox")
	r, err := cli(ctx, b.docker, append(argv, args...), stdin, nil, nil)
	if err != nil {
		return r, err
	}
	if r.code != 0 && gone.MatchString(r.stderr.String()) {
		return r, ports.SandboxGone(b.id, "")
	}
	return r, nil
}

func (b *box) RunCommand(ctx context.Context, command string, opts ports.RunCommandOptions) (ports.CommandResult, error) {
	limit := opts.Timeout
	if limit <= 0 {
		limit = ports.DefaultCommandTimeout
	}
	cwd := opts.Cwd
	if cwd == "" {
		cwd = "."
	}
	args := []string{"exec", "-w", b.path(cwd)}
	for k, v := range opts.Env {
		args = append(args, "-e", k+"="+v)
	}
	cmd, secs := command, sandboxsh.TimeoutSeconds(limit)
	onOut, onErr := opts.OnStdout, opts.OnStderr
	if opts.Background {
		cmd, secs, onOut, onErr = sandboxsh.BackgroundCommand(command), 0, nil, nil
	}
	args = append(append(args, b.id), sandboxsh.RunArgv(cmd, secs, false)...)
	out, errOut := sandboxsh.NewOutput(onOut), sandboxsh.NewOutput(onErr)
	started := time.Now()
	// timeout inside the container stops the command; this is the backstop
	// for an image without it.
	runCtx, cancel := context.WithTimeout(ctx, limit+5*time.Second)
	defer cancel()
	r, err := cli(runCtx, b.docker, args, nil, out, errOut)
	if ctx.Err() != nil {
		return ports.CommandResult{}, ctx.Err()
	}
	killed := runCtx.Err() != nil
	if err != nil && !killed {
		return ports.CommandResult{}, err
	}
	stdout, stderr := out.String(), errOut.String()
	if r.code != 0 && gone.MatchString(stderr) {
		return ports.CommandResult{}, ports.SandboxGone(b.id, "")
	}
	elapsed := time.Since(started)
	if opts.Background {
		return ports.CommandResult{ExitCode: r.code, DurationMs: elapsed.Milliseconds()}, nil
	}
	timedOut := killed || sandboxsh.LooksTimedOut(r.code, elapsed, limit)
	code := r.code
	if timedOut {
		code = ports.TimedOutExitCode
	}
	return ports.CommandResult{
		Stdout: stdout, Stderr: stderr, ExitCode: code, DurationMs: elapsed.Milliseconds(),
		TimedOut: timedOut, Truncated: out.Truncated || errOut.Truncated,
	}, nil
}

func (b *box) GetURL(ctx context.Context, port int, protocol string) (string, error) {
	r, err := cli(ctx, b.docker, []string{"port", b.id, fmt.Sprintf("%d/tcp", port)}, nil, nil, nil)
	if err != nil {
		return "", err
	}
	if r.code != 0 && gone.MatchString(r.stderr.String()) {
		return "", ports.SandboxGone(b.id, "")
	}
	hostPort := strings.SplitN(strings.TrimSpace(r.stdout.String()), "\n", 2)[0]
	if r.code != 0 || hostPort == "" {
		return "", ports.SandboxUnsupported("docker", fmt.Sprintf("a URL for port %d (publish it with the ports option)", port))
	}
	if protocol == "" {
		protocol = "http"
	}
	return protocol + "://" + strings.Replace(hostPort, "0.0.0.0", "127.0.0.1", 1), nil
}

func (b *box) GetInfo(ctx context.Context) (ports.SandboxInfo, error) {
	r, err := cli(ctx, b.docker, []string{"inspect", "-f", "{{.Created}}\t{{.State.Status}}\t{{json .Config.Labels}}", b.id}, nil, nil, nil)
	if err != nil {
		return ports.SandboxInfo{}, err
	}
	if r.code != 0 {
		return ports.SandboxInfo{}, ports.SandboxGone(b.id, "")
	}
	fields := strings.SplitN(strings.TrimSpace(r.stdout.String()), "\t", 3)
	for len(fields) < 3 {
		fields = append(fields, "")
	}
	created, _ := time.Parse(time.RFC3339Nano, fields[0])
	var labels map[string]string
	_ = json.Unmarshal([]byte(fields[2]), &labels)
	md := map[string]string{}
	for k, v := range labels {
		if strings.HasPrefix(k, "agentenkit.") {
			md[strings.TrimPrefix(k, "agentenkit.")] = v
		}
	}
	status := "error"
	switch fields[1] {
	case "running":
		status = "running"
	case "exited":
		status = "stopped"
	}
	info := ports.SandboxInfo{ID: b.id, Provider: "docker", Status: status, CreatedAt: created, Workdir: b.workdir, Metadata: md}
	d, err := b.script(ctx, "cat "+deadlineFile, nil, nil)
	if err != nil {
		return info, err
	}
	if secs, err := strconv.ParseInt(strings.TrimSpace(d.stdout.String()), 10, 64); err == nil && secs > 0 {
		info.ExpiresAt = time.Unix(secs, 0)
	}
	return info, nil
}

func (b *box) SetTimeout(ctx context.Context, d time.Duration) error {
	deadline := strconv.FormatInt(time.Now().Add(d).Unix(), 10)
	r, err := b.script(ctx, `echo "$1" > `+deadlineFile, []string{deadline}, nil)
	if err != nil {
		return err
	}
	if r.code != 0 {
		return fmt.Errorf("docker: could not set the sandbox's time: %s", strings.TrimSpace(r.stderr.String()))
	}
	return nil
}

func (b *box) Destroy(ctx context.Context) error {
	r, err := cli(ctx, b.docker, []string{"rm", "-f", b.id}, nil, nil, nil)
	if err != nil {
		return err
	}
	if r.code != 0 && !gone.MatchString(r.stderr.String()) {
		return fmt.Errorf("docker: could not remove %s: %s", b.id, strings.TrimSpace(r.stderr.String()))
	}
	return nil
}

type files struct{ b *box }

func (f *files) run(ctx context.Context, script, p string, stdin []byte) (result, error) {
	r, err := f.b.script(ctx, script, []string{f.b.path(p)}, stdin)
	if err != nil {
		return r, err
	}
	if r.code == sandboxsh.NotFoundExit {
		return r, ports.SandboxFileNotFound(p)
	}
	if r.code != 0 {
		msg := strings.TrimSpace(r.stderr.String())
		if msg == "" {
			msg = fmt.Sprintf("exit %d", r.code)
		}
		return r, errors.New("docker: " + msg)
	}
	return r, nil
}

func (f *files) ReadFile(ctx context.Context, p string) (string, error) {
	b, err := f.ReadFileBytes(ctx, p)
	return string(b), err
}

func (f *files) ReadFileBytes(ctx context.Context, p string) ([]byte, error) {
	r, err := f.run(ctx, sandboxsh.ReadScript, p, nil)
	if err != nil {
		return nil, err
	}
	return r.stdout.Bytes(), nil
}

func (f *files) WriteFile(ctx context.Context, p string, content []byte) error {
	if content == nil {
		content = []byte{}
	}
	_, err := f.run(ctx, sandboxsh.WriteScript, p, content)
	return err
}

func (f *files) ReadDir(ctx context.Context, p string) ([]ports.FileEntry, error) {
	r, err := f.run(ctx, sandboxsh.ListScript, p, nil)
	if err != nil {
		return nil, err
	}
	return sandboxsh.ParseListing(r.stdout.String()), nil
}

func (f *files) Mkdir(ctx context.Context, p string) error {
	_, err := f.run(ctx, sandboxsh.MkdirScript, p, nil)
	return err
}

func (f *files) Exists(ctx context.Context, p string) (bool, error) {
	r, err := f.b.script(ctx, sandboxsh.ExistsScript, []string{f.b.path(p)}, nil)
	if err != nil {
		return false, err
	}
	return r.code == 0, nil
}

func (f *files) Remove(ctx context.Context, p string) error {
	_, err := f.run(ctx, sandboxsh.RemoveScript, p, nil)
	return err
}

type result struct {
	code           int
	stdout, stderr *bytes.Buffer
}

// cli runs the docker CLI. With out and errOut the output goes there as it
// comes; otherwise it is collected whole. A run cut short by ctx returns
// what it had with ctx's error.
func cli(ctx context.Context, docker string, args []string, stdin []byte, out, errOut *sandboxsh.Output) (result, error) {
	r := result{stdout: &bytes.Buffer{}, stderr: &bytes.Buffer{}}
	cmd := exec.CommandContext(ctx, docker, args...)
	cmd.WaitDelay = 2 * time.Second
	if stdin != nil {
		cmd.Stdin = bytes.NewReader(stdin)
	}
	cmd.Stdout, cmd.Stderr = r.stdout, r.stderr
	if out != nil {
		cmd.Stdout = out
	}
	if errOut != nil {
		cmd.Stderr = errOut
	}
	err := cmd.Run()
	var exitErr *exec.ExitError
	switch {
	case ctx.Err() != nil:
		r.code = -1
		return r, ctx.Err()
	case errors.As(err, &exitErr):
		r.code = exitErr.ExitCode()
		return r, nil
	case err != nil:
		return r, fmt.Errorf("docker: could not run %s: %w", docker, err)
	}
	return r, nil
}
